Added nginx as an optional server.

ajgon opened the pull request:
Why? For the glory of satan of course...
Also, many people tends to use their PIs for multiple purposes. As mail
servers, webpage hosting etc. Usually in that cases nginx is used as far
more powerful alternative to lighttpd, and fair more lighter than
apache. This patch allows them to unleash the power of pi-hole without
too much hacking.

Also, I'm the one of those people, and needed it ;).

dschaper commented:
Branched to https://github.com/pi-hole/pi-hole/tree/Feature/ajgon-nginx. Please push to this new branch. Thanks!

dschaper commented:
I seem to be getting a permissions error with the configuration:

```
2016/01/29 00:06:07 [error] 20560#0: *1 directory index of "/var/www/html/" is forbidden, client: fd7a:1df3:497f:0:418d:dde6:451e:d02, server: , request: "GET / HTTP/1.1", host: "raspberrypi"
2016/01/29 00:06:07 [error] 20560#0: *1 open() "/var/www/html/favicon.ico" failed (2: No such file or directory), client: fd7a:1df3:497f:0:418d:dde6:451e:d02, server: , request: "GET /favicon.ico HTTP/1.1", host: "raspberrypi", referrer: "http://raspberrypi/"
2016/01/29 00:07:50 [error] 20560#0: *1 directory index of "/var/www/html/" is forbidden, client: fd7a:1df3:497f:0:418d:dde6:451e:d02, server: , request: "GET / HTTP/1.1", host: "raspberrypi"
2016/01/29 00:07:52 [error] 20560#0: *1 directory index of "/var/www/html/" is forbidden, client: fd7a:1df3:497f:0:418d:dde6:451e:d02, server: , request: "GET / HTTP/1.1", host: "raspberrypi"
2016/01/29 00:07:52 [error] 20560#0: *1 open() "/var/www/html/favicon.ico" failed (2: No such file or directory), client: fd7a:1df3:497f:0:418d:dde6:451e:d02, server: , request: "GET /favicon.ico HTTP/1.1", host: "raspberrypi", referrer: "http://raspberrypi/"
```

```
pi@raspberrypi:/var/www/html $ ls -lac
total 24
drwxrwxr-x 4 www-data www-data 4096 Jan 29 00:03 .
drwxr-xr-x 3 root     root     4096 Jan 28 09:30 ..
drwxr-xr-x 6 root     root     4096 Jan 28 09:32 admin
-rw-r--r-- 1 root     root     3373 Jan 28 09:32 index.lighttpd.orig
-rw-r--r-- 1 root     root      867 Jan 29 00:02 index.nginx-debian.html
drwxr-xr-x 2 root     root     4096 Jan 28 09:32 pihole
```

But the admin interface works great, we get php-fpm capabilities and the `x-header` is still there... One of the big reasons @jacobsalmela stayed with lighttpd was the ability to use custom headers for debugging purposes, and this seems to have it. I would suggest to Jacob that we consider moving from `lighttpd` to `nginx` for the default configuration, I really think it's better in the long and short run.

ajgon commented:
Hey, glad you like it :)

The permission problem occurs, due to fact that there is no index file in `/var/www/html` by default. Without it, nginx falls back to listing the directory which is forbidden. I can think of three solutions of the problem:
1. put some dummy (even empty) index.html file
2. fall back to default index.nginx-debian.html file (I believe current lighttpd setup does that with its index.lighttpd.html)
3. allow directory indexing (bad idea IMHO).

Regarding the X-Header - if it's that important, I have to do some fixes as well. I'm not sure which nginx version is by default in Raspbian, but prior to version 1.7.4 nginx doesn't support setting HTTP Headers on 4xx and 5xx requests (unless you precompile an extra module). Which means, there won't be any extra headers in all DNS-passed requests (since they are 404s). But fear not, I'll figure something out later today :)

Mcat12 commented:
I think the index.nginx-debian.html is the best option currently, as that's what lighttpd does.

ajgon commented:
I fixed the headers and directoryindex issue, and moved it to the new pull-request (which will merge to the asked branch): https://github.com/pi-hole/pi-hole/pull/271
